### PR TITLE
[Cosmos] add logging for user database account

### PR DESCRIPTION
**Pull request by simorenoh**
This PR makes it so we ensure we log the user's database account settings for us to have available in the logs.

Relevant log sections look like this:
image

Progress towards closing https://github.com/Azure/azure-sdk-for-go/issues/18986.

### CODE REVIEW

File: sdk/data/azcosmos/cosmos_global_endpoint_manager.go [hunk 1]
```diff
@@ -143,15 +146,23 @@ func (gem *globalEndpointManager) GetAccountProperties(ctx context.Context) (acc
 		if err != nil {
 			return accountProperties{}, fmt.Errorf("failed to parse account properties: %v", err)
 		}
+		gem.once.Do(func() {
```
Comment (ealsur) [inline sdk/data/azcosmos/cosmos_global_endpoint_manager.go:149]: Why once? Wouldn't it be beneficial to log whenever anything changes?

File: sdk/data/azcosmos/cosmos_global_endpoint_manager.go [hunk 2]
```diff
@@ -143,15 +146,23 @@ func (gem *globalEndpointManager) GetAccountProperties(ctx context.Context) (acc
...
+		})
 		return properties, nil
 	}
 
 	return accountProperties{}, azruntime.NewResponseErrorWithErrorCode(azResponse, azResponse.Status)
 }
 
 func newAccountProperties(azResponse *http.Response) (accountProperties, error) {
+	var props map[string]interface{}
+	err := azruntime.UnmarshalAsJSON(azResponse, &props)
```
Comment (ealsur) [inline sdk/data/azcosmos/cosmos_global_endpoint_manager.go:160]: Why is this needed?

File: sdk/data/azcosmos/cosmos_global_endpoint_manager.go (hunk 1 above)
Comment (simorenoh) [reply sdk/data/azcosmos/cosmos_global_endpoint_manager.go:149]: we run update every 5 minutes so it's not necessarily every time something changes - but yeah that's why I raised this question in the description

File: sdk/data/azcosmos/cosmos_global_endpoint_manager.go [hunk 3]
```diff
@@ -143,15 +146,23 @@ func (gem *globalEndpointManager) GetAccountProperties(ctx context.Context) (acc
 		if err != nil {
 			return accountProperties{}, fmt.Errorf("failed to parse account properties: %v", err)
 		}
+		gem.once.Do(func() {
+			log.Write(azlog.EventResponse, "===== Database Account Information:\n"+properties.String()+"\n=====\n")
```
[the remark this answers is not in the document]
Comment (simorenoh) [reply sdk/data/azcosmos/cosmos_global_endpoint_manager.go:150]: we can technically just make our own event type, personally I don't think it looks confusing but that may just be me - the event is very neatly outlined from the rest of the logs:
image

File: sdk/data/azcosmos/cosmos_global_endpoint_manager.go (hunk 1 above)
Comment (ealsur) [reply sdk/data/azcosmos/cosmos_global_endpoint_manager.go:149]: It might be relevant to log every time it changes, and if we cannot tell when that is, at least every time

Comment (simorenoh) [reply sdk/data/azcosmos/cosmos_global_endpoint_manager.go:149]: makes sense, will update this to log every time then

**Comment by simorenoh** [comment]
@ealsur only pending question would be on the formatting - are we ok with just sending the unfiltered account properties or should we do the parsing like shown in the description? difference is this:
image

vs this:
image

File: sdk/data/azcosmos/cosmos_location_cache.go [hunk 4]
```diff
@@ -40,10 +41,21 @@ type accountRegion struct {
...
+}
+
 type accountProperties struct {
-	ReadRegions                  []accountRegion `json:"readableLocations"`
-	WriteRegions                 []accountRegion `json:"writableLocations"`
-	EnableMultipleWriteLocations bool            `json:"enableMultipleWriteLocations"`
+	ReadRegions                  []accountRegion       `json:"readableLocations"`
+	WriteRegions                 []accountRegion       `json:"writableLocations"`
+	EnableMultipleWriteLocations bool                  `json:"enableMultipleWriteLocations"`
+	UserConsistencyPolicy        userConsistencyPolicy `json:"userConsistencyPolicy"`
```
Comment (ealsur) [inline sdk/data/azcosmos/cosmos_location_cache.go:52]:
```suggestion
	AccountConsistency accountConsistency `json:"userConsistencyPolicy"`
```

File: sdk/data/azcosmos/cosmos_location_cache.go [hunk 5]
```diff
@@ -40,10 +41,21 @@ type accountRegion struct {
...
+	ContinuousBackupEnabled      bool                  `json:"continousBackupEnabled"`
```
Comment (ealsur) [inline sdk/data/azcosmos/cosmos_location_cache.go:53]: This property is not relevant to the client SDK

File: sdk/data/azcosmos/cosmos_location_cache.go [hunk 6]
```diff
@@ -40,10 +41,21 @@ type accountRegion struct {
...
+}
+
+func (accountProps accountProperties) String() string {
+	return fmt.Sprintf("Read Regions: %v\nWrite Regions: %v\nUse Multiple Write Locations: %v\nDefault Consistency Level: %v\nContinous Backup Enabled: %v",
```
Comment (ealsur) [inline sdk/data/azcosmos/cosmos_location_cache.go:57]:
```suggestion
	return fmt.Sprintf("Read Regions: %v\nWrite Regions: %v\nMulti-region writes : %v\nDefault Consistency Level: %v",
```

Portal uses "Multi-region writes"
image

Might as well be consistent

File: sdk/data/azcosmos/cosmos_location_cache.go [hunk 7]
```diff
@@ -40,10 +41,21 @@ type accountRegion struct {
...
+
 type accountProperties struct {
-	ReadRegions                  []accountRegion `json:"readableLocations"`
-	WriteRegions                 []accountRegion `json:"writableLocations"`
-	EnableMultipleWriteLocations bool            `json:"enableMultipleWriteLocations"`
+	ReadRegions                  []accountRegion       `json:"readableLocations"`
+	WriteRegions                 []accountRegion       `json:"writableLocations"`
+	EnableMultipleWriteLocations bool                  `json:"enableMultipleWriteLocations"`
+	AccountConsistency accountConsistency `json:"userConsistencyPolicy"`
+	ContinuousBackupEnabled      bool                  `json:"continousBackupEnabled"`
```
Comment (simorenoh) [inline sdk/data/azcosmos/cosmos_location_cache.go:53]:
```suggestion
```

File: sdk/data/azcosmos/cosmos_location_cache.go [hunk 8]
```diff
@@ -40,10 +41,20 @@ type accountRegion struct {
...
-	EnableMultipleWriteLocations bool            `json:"enableMultipleWriteLocations"`
+	ReadRegions                  []accountRegion       `json:"readableLocations"`
+	WriteRegions                 []accountRegion       `json:"writableLocations"`
+	EnableMultipleWriteLocations bool                  `json:"enableMultipleWriteLocations"`
+	AccountConsistency accountConsistency `json:"userConsistencyPolicy"`
+}
+
+func (accountProps accountProperties) String() string {
+	return fmt.Sprintf("Read Regions: %v\nWrite Regions: %v\nMulti-region writes: %v\nDefault Consistency Level: %v",
+		accountProps.ReadRegions, accountProps.WriteRegions, accountProps.EnableMultipleWriteLocations, accountProps.UserConsistencyPolicy.DefaultConsistencyLevel, accountProps.ContinuousBackupEnabled)
```
Comment (simorenoh) [inline sdk/data/azcosmos/cosmos_location_cache.go:57]:
```suggestion
		accountProps.ReadRegions, accountProps.WriteRegions, accountProps.EnableMultipleWriteLocations, accountProps.AccountConsistency.DefaultConsistencyLevel, accountProps.ContinuousBackupEnabled)
```

File: sdk/data/azcosmos/emulator_cosmos_global_endpoint_manager_test.go [hunk 9]
```diff
@@ -30,6 +30,7 @@ func TestGlobalEndpointManagerEmulator(t *testing.T) {
 		ReadRegions:                  []accountRegion{emulatorRegion},
 		WriteRegions:                 []accountRegion{emulatorRegion},
 		EnableMultipleWriteLocations: false,
+		UserConsistencyPolicy:        userConsistencyPolicy{"Strong"},
```
Comment (simorenoh) [inline sdk/data/azcosmos/emulator_cosmos_global_endpoint_manager_test.go:33]:
```suggestion
		AccountConsistency:        userConsistencyPolicy{"Strong"},
```

Comment (ealsur) [reply sdk/data/azcosmos/emulator_cosmos_global_endpoint_manager_test.go]: Coding through suggestions :D

File: sdk/data/azcosmos/cosmos_location_cache.go [hunk 10]
```diff
@@ -40,10 +41,20 @@ type accountRegion struct {
...
-	EnableMultipleWriteLocations bool            `json:"enableMultipleWriteLocations"`
+	ReadRegions                  []accountRegion       `json:"readableLocations"`
+	WriteRegions                 []accountRegion       `json:"writableLocations"`
+	EnableMultipleWriteLocations bool                  `json:"enableMultipleWriteLocations"`
+	AccountConsistency accountConsistency `json:"userConsistencyPolicy"`
+}
+
+func (accountProps accountProperties) String() string {
+	return fmt.Sprintf("Read Regions: %v\nWrite Regions: %v\nMulti-region writes: %v\nDefault Consistency Level: %v",
+		accountProps.ReadRegions, accountProps.WriteRegions, accountProps.EnableMultipleWriteLocations, accountProps.AccountConsistency.DefaultConsistencyLevel, accountProps.ContinuousBackupEnabled)
```
Comment (simorenoh) [inline sdk/data/azcosmos/cosmos_location_cache.go:57]:
```suggestion
func (accountProps accountProperties) String() string {
	return fmt.Sprintf("Read Regions: %v\nWrite Regions: %v\nMulti-region writes: %v\nDefault Consistency Level: %v",
		accountProps.ReadRegions, accountProps.WriteRegions, accountProps.EnableMultipleWriteLocations, accountProps.AccountConsistency.DefaultConsistencyLevel)
```

File: sdk/data/azcosmos/cosmos_location_cache.go [hunk 11]
```diff
@@ -40,10 +41,20 @@ type accountRegion struct {
...
+
 type accountProperties struct {
-	ReadRegions                  []accountRegion `json:"readableLocations"`
-	WriteRegions                 []accountRegion `json:"writableLocations"`
-	EnableMultipleWriteLocations bool            `json:"enableMultipleWriteLocations"`
+	ReadRegions                  []accountRegion       `json:"readableLocations"`
+	WriteRegions                 []accountRegion       `json:"writableLocations"`
+	EnableMultipleWriteLocations bool                  `json:"enableMultipleWriteLocations"`
+	AccountConsistency accountConsistency `json:"userConsistencyPolicy"`
+}
```
Comment (simorenoh) [inline sdk/data/azcosmos/cosmos_location_cache.go:53]:
```suggestion
	ReadRegions                  []accountRegion       `json:"readableLocations"`
	WriteRegions                 []accountRegion       `json:"writableLocations"`
	EnableMultipleWriteLocations bool                  `json:"enableMultipleWriteLocations"`
	AccountConsistency userConsistencyPolicy `json:"userConsistencyPolicy"`
}
```

File: sdk/data/azcosmos/emulator_cosmos_global_endpoint_manager_test.go [hunk 12]
```diff
@@ -30,6 +30,7 @@ func TestGlobalEndpointManagerEmulator(t *testing.T) {
 		ReadRegions:                  []accountRegion{emulatorRegion},
 		WriteRegions:                 []accountRegion{emulatorRegion},
 		EnableMultipleWriteLocations: false,
+		AccountConsistency:           userConsistencyPolicy{"Strong"},
```
Comment (ealsur) [inline sdk/data/azcosmos/emulator_cosmos_global_endpoint_manager_test.go:33]: nit: Are there any tests where we are checking emitted logs that need to be updated?

**Review by ealsur — APPROVED** [review]

File: sdk/data/azcosmos/emulator_cosmos_global_endpoint_manager_test.go (hunk 12 above)
Comment (simorenoh) [reply sdk/data/azcosmos/emulator_cosmos_global_endpoint_manager_test.go:33]: nope - just the one test I had to update on verifying account settings for the emulator
